### DISCUSSION TITLE
How can I decompile C# code?

### DISCUSSION THREAD

**Discussion by Shervan360** [icsharpcode/ILSpy#2892] [General]
Hello,

I am seeing a tutorial video and the instructor decompiled C# code with ILSpy. I want to do the same but my result is different. 
Please see the two attached photos.  The first photo is mine, and the second is by the teacher.
ILSpy version 8.0.0.7246-preview3

![Screenshot 2023-02-05 174150](https://user-images.githubusercontent.com/60143155/216863473-fe9d73e7-8a61-4c29-942b-d8c7f21b12ff.png)
![Screenshot 2023-02-05 174138](https://user-images.githubusercontent.com/60143155/216863484-1d6e8563-4c5b-48b1-98c1-db55deeca028.png)

**Comment by Shervan360** [comment]
@siegfriedpammer  Could you please answer me?

**Comment by dgrunwald** [comment]
You are using a different language setting than your teacher.
ILSpy will decompile low-level constructs like `ValueTuple` into the corresponding higher-level C# features (like tuple types) if possible. But you can disable this by selecting a lower language version. For example, tuple types were introduced in C# 7.
So if you select C# 7 or newer you will see `(Guid, IEnumerable<Item>)` and if you select C# 6 or older, you will see `ValueTuple<Guid, IEnumerable<Item>>`. Both variants of the code have exactly the same meaning and result in the same IL code.

**Reply by Shervan360** [reply]
Thank you, Could you please explain more about _"ILSpy will decompile low-level constructs like ValueTuple into the corresponding higher-level C# features (like tuple types) if possible."_ ?
If I wrote tuple type in C# 10, Is C# Microsoft compiler translating it to ValueTuple before translating it to IL?

**Reply by dgrunwald** [reply]
Yes. Tuple types only exist in C#, not in IL; so the C# compiler will translate tuple types to ValueTuple.
ILSpy will then (depending on the selected language version) translate all ValueTuple usages back to tuple types.
